hscontrol/types/config.go

```diff
@@ -274,6 +275,7 @@ func LoadConfig(path string, isFile bool) error {
 	viper.SetDefault("oidc.only_start_if_oidc_is_available", true)
 	viper.SetDefault("oidc.expiry", "180d")
 	viper.SetDefault("oidc.use_expiry_from_token", false)
+	viper.SetDefault("oidc.map_legacy_users", true)
```

Comment:
As I agree that this feature would be problematic, and the best might be to do a clean cut (but that requires us getting it "right" this time), I am thinking of an alternative approach:

No automatic migration, remove this option and all logins will create new users.
Either: 
- a new "merge" command, moving all nodes (and preauthkeys) from one user to another (so the admin can script the merger them selves, knowing the variables of their environment)
- Not add a new command, but show an example of how this can be done with the already existing `headscale nodes move` command.

Feels like this will be messy either way, but this approach is more explicit than implicit.

Reply:
I'm a fan of an explicit approach, but mindful of the overhead if it needs operator intervention.

> No automatic migration, remove this option and all logins will create new users. Either:
> 
> * a new "merge" command, moving all nodes (and preauthkeys) from one user to another (so the admin can script the merger them selves, knowing the variables of their environment)
> 
> * Not add a new command, but show an example of how this can be done with the already existing `headscale nodes move` command.

I think with both of those approaches, the process would be something like:

1. Administrator upgrades Headscale, and announces to users that things will break.

   *This is effectively the start of a service outage for all users.*

2. User attempts to sign in to Headscale with OIDC. Headscale creates a new account, and the user discovers that they are locked out of everything.

   *This is when the user discovers the outage, if they didn't get the administrator's announcement.*

3. Administrator (somehow?) confirms that the new user account created with OIDC belongs to some old account.

4. Administrator runs a command to merge the new account to their old account.

   *This is when the outage is resolved for the user.*

There's some back and forth with the administrator needing to do things for the user – so there's a potential for delays and overheads.

## Alternative

Another way to do this is to have users authenticate *without* OIDC (ie: with their master password) if an account already exists for that email address. That way the linking can happen automatically, but it is checked against "something they know"... rather than trusting the email provided by the IdP implicitly or relying on the contents of `preferred_username`.

Then the migration process for an existing user would be:

1. Administrator upgrades Headscale, opts in to auto-migration.

   *This is effectively the start of a service outage for all users.*

2. User attempts to sign in to Headscale with OIDC.
3. Headscale can't find any account with that `sub`, so falls back to email address matching. It finds an account, which has no `sub` recorded.
4. Headscale prompts the user to either provide the master password for the existing account (to re-establish linking), or to create a new account.
5. User enters the master password for the existing account.
6. Headscale records the provided `sub`, and logs the user in normally.

   *This is when the outage is resolved for the user.*

7. Going forward, user can sign in with OIDC just as they always did, and is matched on `sub`. If details changed, they're automatically updated on log-in.

The benefit of that is it's entirely self-service. But it couldn't handle things like:

* if multiple accounts are associated with the same email address
* if a user's email address changed between the last time they logged in pre-upgrade, and the next time they next logged in post-upgrade

An administrator could also do the migration manually themselves:

1. Administrator dumps a list of users with their validated email address and `sub` from the IdP.
2. Administrator dumps a list of users from Headscale.
3. Administrator creates a list of known valid linkages for both sides. If there's any unlinked entries, they can be followed up separately, and they can also handle things like renamed users.
4. Administrator upgrades Headscale, and does not opt in to auto-migration.

   *This is effectively the start of a service outage for all users.*

5. Administrator runs some command like `headscale user ${username} set-oidc-identifier "${issuer}/${sub}"` for each validated user.

   *This is when the outage is resolved.*

Reply:
Argh, I just realised I put in bunch of commentary about a "master password" which *doesn't* apply to Headscale. That's a Bitwarden/Vaultwarden thing... and I'm getting my apps mixed up. 🙃 

I've updated my previous comment accordingly, but I suspect having a Headscale password may not be a guarantee.

Reply:
Argh (again), I've now also realised that #2020 added the email address field in the first place.

So the only way you could reliably auto-migrate is if `strip_email_domain = false`.